src/lib.rs

```diff
@@ -1452,6 +1452,141 @@ fn collect_runs(
     Ok(out_list)
 }
 
+#[pyfunction]
+#[text_signature = "(graph, special_filter_fn, filter_fn, color_fn)"]
+fn collect_bicolor_runs(
+    py: Python,
+    graph: &digraph::PyDiGraph,
+    special_filter_fn: PyObject,
+    filter_fn: PyObject,
+    color_fn: PyObject,
+) -> PyResult<Vec<Vec<PyObject>>> {
+    let mut pending_list: Vec<Vec<PyObject>> = Vec::new();
+    let mut block_id: Vec<Option<usize>> = Vec::new();
+    let mut block_list: Vec<Vec<PyObject>> = Vec::new();
+
+    let special_filter_node = |node: &PyObject| -> PyResult<bool> {
+        let res = special_filter_fn.call1(py, (node,))?;
+        res.extract(py)
+    };
+
+    let filter_node = |node: &PyObject| -> PyResult<bool> {
+        let res = filter_fn.call1(py, (node,))?;
+        res.extract(py)
+    };
+
+    let color_edge = |edge: &PyObject| -> PyResult<i32> {
+        let res = color_fn.call1(py, (edge,))?;
+        res.extract(py)
+    };
+
+    let nodes = match algo::toposort(graph, None) {
+        Ok(nodes) => nodes,
+        Err(_err) => {
+            return Err(DAGHasCycle::new_err("Sort encountered a cycle"))
+        }
+    };
+
+    // Utility for ensuring pending_list has the color index
+    macro_rules! ensure_vector_has_index {
+        ($pending_list: expr, $block_id: expr, $color: expr) => {
+            if $color >= $pending_list.len() {
+                $pending_list.resize($color + 1, Vec::new());
+                $block_id.resize($color + 1, None);
+            }
+        };
+    }
+
+    for node in nodes {
+        if special_filter_node(&graph.graph[node])? {
```

Comment:
why do we need this `special_filter_node` function? isn't `filter_node` enough?

Reply:
`special_filter_node` + topological sort  is the equivalent of [this](https://github.com/Qiskit/qiskit-terra/pull/6534/files#diff-4be18ab968355b22882ab1c3b52904b231c102a68a77d8ea3c5a2066f0511b14R41). That happens because [`topological_op_nodes`](https://github.com/Qiskit/qiskit-terra/blob/31a24c1a6fbbe62d7c1c568e405cde92f87fc0b7/qiskit/dagcircuit/dagcircuit.py#L945) filter only ops. To replicate the exact behavior I opted to add `speecial_filter_fn` which will receive the filter `nd.type == 'op'`

Reply:
With that said, I think it could be replaced only with `filter_fn` but performance will be impacted. The `special_filter_fn` is cheap to execute and prunes many loops in edges that call a more expensive `color_fn`

Reply:
i guess, in most dagcircuits `op` nodes will be way more than `in` + `out` nodes and it seems to me that altough `special_filter_fn` prunes some loops, we pay an overhead for calling it when most of the time it returns `true`.

Reply:
I opted to simplify this with `filter_fn` using ternary logic. If the result is `None`, then we simply ignore the node as it is not an operation. Otherwise, we get `True` or `False` and can handle the situation accordingly. This leads to a single Python call